contracts/src/investment/UniV2StakeFarm.sol

```diff
@@ -0,0 +1,365 @@
+    emit ControllerChanged(newController);
+  }
+
+  function notifyRewardAmount(uint256 reward)
+    external
+    override
+    onlyController
+    updateReward(address(0))
+  {
+    if (block.timestamp >= periodFinish) {
+      rewardRate = reward.div(rewardsDuration);
+    } else {
+      uint256 remaining = periodFinish.sub(block.timestamp);
+      uint256 leftover = remaining.mul(rewardRate);
+      rewardRate = reward.add(leftover).div(rewardsDuration);
+    }
+    availableRewards = availableRewards.add(reward);
+
+    // Ensure the provided reward amount is not more than the balance in the contract.
+    // This keeps the reward rate in the right range, preventing overflows due to
+    // very high values of rewardRate in the earned and rewardsPerToken functions;
+    // Reward + leftover must be less than 2^256 / 10^18 to avoid overflow.
+    require(
+      rewardRate <= availableRewards.div(rewardsDuration),
+      'Provided reward too high'
+    );
+
+    lastUpdateTime = block.timestamp;
+    periodFinish = block.timestamp.add(rewardsDuration);
+    emit RewardAdded(reward);
+  }
+
+  // we don't have any rebalancing here
+  function rebalance() external override onlyController {}
+
+  // Added to support recovering LP Rewards from other systems to be distributed to holders
+  function recoverERC20(address tokenAddress, uint256 tokenAmount)
+    external
+    onlyOwner
+  {
+    // Cannot recover the staking token or the rewards token
+    require(
+      tokenAddress != address(stakingToken),
+      'pool tokens not recoverable'
+    );
+    IERC20(tokenAddress).safeTransfer(owner(), tokenAmount);
+    emit Recovered(tokenAddress, tokenAmount);
+  }
+
+  function setRewardsDuration(uint256 _rewardsDuration)
+    external
+    override
+    onlyOwner
+  {
+    require(
+      periodFinish == 0 || block.timestamp > periodFinish,
+      'reward period not finished'
+    );
+    rewardsDuration = _rewardsDuration;
+    emit RewardsDurationUpdated(rewardsDuration);
+  }
+
+  /* ========== PRIVATE ========== */
+
+  function _ethAmount(uint256 amountToken) private view returns (uint256) {
+    (uint112 reserve0, uint112 reserve1, ) = stakingToken.getReserves();
+    // WETH is token 1
+    if ((pairDirection & 1) != 0) reserve0 = reserve1;
+
+    return (uint256(reserve0).mul(amountToken)).div(stakingToken.totalSupply());
+  }
+
+  /**
+   * @dev returns the reserves in order: ETH -> Token, ETH/stable
+   */
+  function _getTokenUiData()
+    internal
+    view
+    returns (
+      uint112,
+      uint112,
+      uint256
+    )
+  {
+    (uint112 reserve0, uint112 reserve1, ) = stakingToken.getReserves();
+    (uint112 reserve0R, uint112 reserve1R, ) =
+      address(route) != address(0) ? route.getReserves() : (1, 1, 0);
+
+    uint112 swap;
+    // WETH is token 1, swap
+    if ((pairDirection & 1) != 0) {
+      swap = reserve0;
+      reserve0 = reserve1;
+      reserve1 = swap;
+    }
+    // WETH is token 1, swap
```

Comment:
Should this be `// WETH is token 2, swap`?

Reply:
they have zero based naming, in a uniswap pair you have token0 and token1, I'll remove the space